base/reduce.jl

```diff
@@ -1246,3 +1264,102 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
     end
     return n
 end
+
+# simd optimization for min/max related reduction
+# general fallback
+_fast(op, x, y) = op(x, y)
+_fast(op) = (x, y) -> _fast(op, x, y)
+
+# The following are used in optimized `mapreduce_impl` for IEEEFloat.
+_fast(::typeof(min), x::T, y::T) where {T<:IEEEFloat} = ifelse(signbit(x-y), x, y)
+_fast(::typeof(max), x::T, y::T) where {T<:IEEEFloat} = ifelse(signbit(y-x), x, y)
+
+# Some help function for better vectorization
+# NaN check
+_anynan(x, y) = isnan(x) | isnan(y)
+_anynan(x::NTuple{N}) where {N} = reduce(|, map(_anynan, x[1:N>>1], x[1+N>>1:N]))
+# split then reduce for inputs whose length > 4
+_half_reduce(op, x::NTuple{N}) where {N} = _half_reduce(op, map(op, x[1:N>>1], x[1+N>>1:N]))
+_half_reduce(op, x::NTuple{4}) = reduce(op, x)
+
+function mapreduce_impl(f, op::Union{typeof(max),typeof(min)},
+                        A::AbstractArrayOrBroadcasted, first::Int, last::Int)
+    @inline elf(i::Int) = @inbounds f(A[i])
+    Eltype = _return_type(elf, Tuple{Int})
+    # Limit this optimization to IEEEFloat. (general fallback is 2x faster for Integer)
+    Eltype <: IEEEFloat ||
+        return mapreduce_impl(f, op, A, first, last, pairwise_blocksize(f, op))
```

Comment:
Using the inference API (`_return_type`) like this should probably ensure that the result is bit-wise equal. The "SIMD friendly" implementation below is requiring the communitativity of `min`/`max` but the generic `mapreduce` does not rely on commutativity. This difference can be observed when the user checks the NaN with `===`. If you want to do this, I think using `eltype` seems to be the only reliable choice.

Also, IIUC, this PR's main purpose is to fix the bugs. I personally think separating optimization-only PR is a better approach.

Reply:
I guess replacing `Eltype <: IEEEFloat` with `isconcretetype(Eltype ) && Eltype <: IEEEFloat` is enough?
It should also solve the following concern.

Reply:
Seperating seems OK to me, will remove them in this PR.

Reply:
No, I don't think it's enough. A subtle change in user code can change the quality of inference of `elf` while preserving the input-output mapping of `elf`. It can happen even if `elf` always returns (say) a `Float64` at run-time.

Reply:
Do you mean something like `f(x::Float64) = isnan(x) ? 0 : x` while inputs has no NaN?
I guess we can give up optimizion in this case.

Reply:
I have open #43725 for early review for performance change.
Further discussion could continue there.